### ISSUE TITLE
router

### ISSUE THREAD

**Issue by lega911** [lega911/angular-light#137]
https://gist.github.com/fend25/3619d6d730039c30a34d

**Comment by fend25** [comment]
https://jsfiddle.net/fend25/e1LLu72c/
better and preferabe variant now,

guys, feel free to comment)

**Comment by dev-rke** [comment]
Haha, thought yesterday how great it would be having some views bound to routes. :D 

A good basic router implementation would be this:
http://krasimirtsonev.com/blog/article/A-modern-JavaScript-router-in-100-lines-history-api-pushState-hash-url

My approach would use regexes as parameter of a directive:

``` html
<div al-router-view="/main/?">
  <a href="/user/1">User 1</a>
</div>
<div al-router-view="/user/:id/?">
  <a href="/main">Back to overview</a>
</div>
```

This would allow to define the urls directly in the template, no additional code would be needed for basic features. But there should definitivly be an API to do redirects, handle url errors etc.

**Comment by fend25** [comment]
Oleg's reference implementation (ancient one)
http://plnkr.co/edit/QYb53NUnnhIezI5n0JxB?p=preview

**Comment by dev-rke** [comment]
Ok, i just build an absolute simple, quick n dirty views implementation:
https://jsfiddle.net/r9t3ropc/

Some explanation: the al-router-view directive could be extended to watch the expression.
So one could define the url structure in javascript or the url of a view could be changed:

``` html
<div al-router-view="myVar.main">Main</div>
<div al-router-view="myVar.user">User</div>
```

``` javascript
scope.myVar = {
  main: '/main',
  user: '/user'
}
```

Comments are welcome.

**Comment by rumkin** [comment]
I think it should be separated into two different API's:
- Router is for router events and state handling
- View rendering route data into template and managing scope.

The first should have unified interface to change current location and update/rewrite state. Also it should be presented in root scope and should broadcast events. Events could be stopped to prevent page reloading when there is unsaved changes or so.

View should allow to use functions as route value. This functions return async operations to load template and resource specified by route. This is needed to render state dependent routes. By default view listening scope's `$routeChanged` event. But changes source could be configurable to use exact router instance.

``` html
<al-app>
    <al-view al-view.routes = "routes"/>
</al-app>
```

``` js
alight.bootstrap('al-app', {
    routes : {
        // Simple template route
        '/': 'index.html',

        // State dependent route
        '/home': function(route, scope){
            if (scope.user) {
                return {
                    scope: {
                        user: user
                    },
                    template: '/users/home.html'
                }
            } else {
                return '/sign-in.html'
            }
        },
        // Async route
        '/users/:id' : route => {
            var user = users.get(route.id);

            var promise;

            if (! user) {
                promise = loadUser(route.id).then(user => {
                    if (! user) {
                        // User not found, show 404 error page
                        return '/errors/404.html';
                    } else {
                        // User found, use user page template
                        return {
                            scope: {
                                user
                            },
                            template: '/users/item.html'
                        }
                    }
                });
            } else {
                promise = Promise.resolve(user);
            }

            return promise;
        }
    }
});
```

Also view could has `loading` view with spinner to indicate intermediate state while resolving promise.

**Comment by lega911** [comment]
@rumkin Can I use common header and footer for "/users/*" in your approach.
In this example by @fend25 I can do this: https://jsfiddle.net/fend25/e1LLu72c/

> Events could be stopped to prevent page reloading when there is unsaved changes or so.

It's `@onLoseFocus="showPreventWindow()"` in the example.

`@onSwitchTo="onSwtichToUser()"` could be something like this:

``` js
scope.onSwtichToUser = route => {
            var user = users.get(route.id).then

            var promise;

            if (! user) {
                promise = loadUser(route.id).then(user => {
                    if (! user) {
                        // User not found, show 404 error page
                        return '/errors/404.html';
                    } else {
                        // User found, use user page template
                        return {
                            scope: {
                                user
                            },
                            template: '/users/item.html'
                        }
                    }
                });
            } else {
                promise = Promise.resolve(user);
            }

            return promise;
}
```

What do you think? "Configuring" in html can be easier.

**Comment by rumkin** [comment]
@lega911 Yep, it could be done with several view elements. Like this:

``` html
<al-view al-view-routes="routes.head">
<al-view al-view-routes="routes.body">
<al-view al-view-routes="routes.footer">
```

It's not so elegant as HTML configuration but it simply to manage from code. I think configuring in HTML is the most proper way to do this for MVVM app. But it should be also configurable and controllable from code.

`@` is not valid HTML attribute name character. Is this some kind of future DSL?

**Comment by rumkin** [comment]
I think it's better to configure router with html and implement fully controllable router and view interfaces with focus on usability.

**Comment by lega911** [comment]
> @ is not valid HTML attribute name character. Is this some kind of future DSL?

Yes, but we can make it like this (with the event directive):
`al-on:on-lose-focus="showPreventWindow()"`
or alias
`@on-lose-focus="showPreventWindow()"`

and "on-lose-focus" is an event (new CustomEvent), which you can catch even with jQuery. https://github.com/lega911/angular-light/issues/123

**Comment by rumkin** [comment]
Maybe it's better to avoid `on` duplication and bind events to target component to avoid mess of events in the future:

```
al-router:on-reload-requested
al-input:on-lost-focus
```

**Comment by lega911** [comment]
Maybe, but custom events can be more flexible.

**Comment by dev-rke** [comment]
May it possible to implement url handling like express.js?
Another point is, that using Views which use "templates" would not be useful, because in this case i could redirect the user to a page which shows the content of this "template".
Views should be used inline, to easily switch between views by a changing url.
I also think it is not useful to bind a view to a scope variable.

My approach uses a global routes object, where regexes are defined.
Using al-route or by adding routes to the object, new routes can be defined.
Also a custom route can be defined in this object, which will be referenced by a given element selector.

``` html
<div id="home"></div>
<div id="players"></div>
<div al-route="/users/:user"></div>
```

``` coffeescript
alight.routes = {
  '/home': (scope) ->
    return '#home'
  '/players/:id': (scope, id) ->
    scope.id = id
    return '#players'
}
alight.d.al.route = (scope, exp, element, env) ->
  alight.routes[exp] = ->
    return element
```

What do you think?

**Comment by lega911** [comment]
> My approach uses a global routes object,

I agree

> `````` html
> div id="home"></div>
> <div id="players"></div>
> <div al-route="/users/:user"></div>```
> ``````

should it work like al-if with dependency on url?

**Comment by lega911** [comment]
What about this one?

``` html
<div>
  <div al-route:group @switch-to="onSwtichToAuth()">
    <div al-route="/login" al-ctrl="App"/>
    <div al-route="/register" al-ctrl="App"/>
  </div>
  <div al-route:group @switch-to="onSwtichToApp()" @lose-focus="showPreventWindow()">
    <Header/>
    <div al-route="/posts" al-ctrl="Posts"/>
    <div al-route="/feed" al-ctrl="Feed"/>
    <div al-route="/user">
      <div al-ctrl="userList"></div>
      <div al-route="/user/:userId" al-ctrl="userItem" @lose-focus="checkIfUserSaved()">
        <div al-if="!user">Loading...</div>
        <div al-if="user" al-include="/templates/userItem.html"></div>
      </div>
    </div>
    <div al-route="*">
      No page 404
    </div>
    <Footer/>
  </div>
</div>
```

when url = "/user/:userId", then controller `al-ctrl="userItem"` is activated, there you can load a user,
a template can be loaded by al-include when a user is loaded `<div al-if="user" al-include="/templates/userItem.html"></div>`,
and you can see this before user is loaded `<div al-if="!user">Loading...</div>`

so, visible html for /user/:userid would be

``` html
<div>
  <div al-route:group @switch-to="onSwtichToApp()" @lose-focus="showPreventWindow()">
    <Header/>
    <div al-route="/user">
      <div al-ctrl="userList"></div>
      <div al-route="/user/:userId" al-ctrl="userItem" @lose-focus="checkIfUserSaved()">
        <div al-if="!user">Loading...</div>
        <div al-if="user" al-include="/templates/userItem.html"></div>
      </div>
    </div>
    <Footer/>
  </div>
</div>
```

visible html for /user would be

``` html
<div>
  <div al-route:group @switch-to="onSwtichToApp()" @lose-focus="showPreventWindow()">
    <Header/>
    <div al-route="/user">
      <div al-ctrl="userList"></div>
    </div>
    <Footer/>
  </div>
</div>
```

visible html for /login would be

``` html
<div>
  <div al-route:group @switch-to="onSwtichToAuth()">
    <div al-route="/login" al-ctrl="App"/>
  </div>
</div>
```

**Comment by rumkin** [comment]
I agree. It looks exciting. And it should to allow to load subroutes with al-include for complex components.

**Comment by lega911** [comment]
@rumkin al-include is deprecated, look at http://angular-light.readthedocs.org/en/latest/directive/html.html , it's more flexible :)

**Comment by rumkin** [comment]
@lega911 I will. But you're using al-include in your example.

Let's focus on modular structure, easy of use and flexibility. And thus it should to use al-html to load subroutes.

**Comment by dev-rke** [comment]
> should it work like al-if with dependency on url?

Yes, you have a pretty good understanding of what i mean. :-)

> al-route:group

I do not understand how this works.
I understand that it groups some requests.
But how is defined that a specific request should be used?
Which request is the first if you are visiting url "/"?

What i also thought about are subroutes:

``` html
<div al-route="/user">
  <div al-route="/:id"></div>
  <div al-route="/details"></div>
</div>
```

Which would be matched when accessing:
/user
/user/:id
/user/details

Maybe it would be great to also handle redirects in the global route object:

``` coffeescript
alight.routes = {
  '/home': '/'
  '/players/:id': (scope, id) ->
    scope.id = id
    return '#players'
}
```

which would redirect "/home" to "/".

**Comment by lega911** [comment]
> al-route:group
> I do not understand how this works.

It's easy, one rule - if there is a route that maches to url, then it should be visible, then all parent DOM should be visible, other route-elements should be hidden.
I changed your example in this style:

``` html
<div al-route>
  <header />
  <div al-route="/profile"></div>
  <div al-route="/user/:id/view"></div>
  <div al-route="/user/details"></div>
  <footer />
</div>
```

for url "/user/details" we will have:

``` html
<div al-route>
  <header />
  <div al-route="/user/details"></div>
  <footer />
</div>
```

In your approach we have to copy-past header and footer for "/profile".
On the other side, your approach works well for loaded html (al-html).

**Comment by dev-rke** [comment]
Hm. This is a bit confusing for me.
The grouping attribute is there to render a specific header an a footer, right?
The content for a route is inside of

``` html
<div al-route="/profile">
  <h1>My user profile</h1>
</div>
```

So when accessing /profile i will see this:

``` html
<header />
<div al-route="/profile">
  <h1>My user profile</h1>
</div>
<footer/>
```

Am i right?

**Comment by lega911** [comment]
> So when accessing /profile i will see this:
> Am i right?

Yes.
Actually for example above we don't need a "group" route:

``` html
<div>
  <header />
  <div al-route="/profile"></div>
  <div al-route="/user/:id/view"></div>
  <div al-route="/user/details"></div>
  <footer />
</div>
```

In a simple, "group route" just holds all child routes, like `al-route="/profile; /user/:id/view; /user/details"`, but it takes them automatically.

**Comment by lega911** [comment]
> The grouping attribute is there to render a specific header an a footer, right?

Not exactly, it renders all DOM, except routes that don't match to URL

**Comment by dev-rke** [comment]
> Not exactly, it renders all DOM, except routes that don't match to URL

I like the approach. :-)

**Comment by lega911** [comment]
Prototype: http://plnkr.co/edit/lZkXWXg5xk1J147nb1jX
embeded: http://run.plnkr.co/plunks/lZkXWXg5xk1J147nb1jX/

with no regex yet.

**Comment by rumkin** [comment]
There is an regex router with params and tail capturing:
https://jsfiddle.net/rumkin/y6xLu1sm/

**Comment by lega911** [comment]
> Here an regex router with params and tail capturing:

I took it for my example

**Comment by lega911** [comment]
You can get arguments using event `al-on.route-to`

``` html
<div al-route="/user/:name/view" @route-to="username=$event.value.name">
```

**Comment by lega911** [comment]
One more example: http://plnkr.co/edit/pQBycumCKE1DJb3oVaPE?p=preview
- you can use the same routes a few times in different places (`al-route="/user/details"` in the example)
- default route

``` html
        <div al-route="*">
          404 - Not found
        </div>
```

**Comment by dev-rke** [comment]
> You can get arguments using event `al-on.route-to`

Why not providing these variables automatically in a scope variable?
E.g. 

``` javascript
scope.$route.name
```

Parsing from URL shouldn't be so difficult: 

``` javascript
"/user/:id/name/:name/view/:action".match(/:\w+/g)
```

**Comment by lega911** [comment]
> Why not providing these variables automatically in a scope variable?

It can give you collisions for routes "/users/:name", "/users/root" they both will be visible on url="/users/root"

**Comment by rumkin** [comment]
@dev-rke It could be done manually with hooks. And sometimes it could cause type mismatch.

I prefer to assign route params to `scope.$route`.

**Comment by dev-rke** [comment]
> It can give you collisions for routes "/users/:name", "/users/root" they both will be visible on url="/users/root"

No, it should not.
When you provide for each route a own scope, e.g.

``` html
<div>
  <header />
  <div al-route="/profile/:profileID">
    {{$route.profileID}}
  </div>
  <div al-route="/user/:id/view">
    {{$route.id}}
  </div>
  <div al-route="/user/details/booking/:id">
    {{$route.id}}
  </div>
  <footer />
</div>
```

Then i see no problems. :-)

**Comment by lega911** [comment]
> No, it should not.
> When you provide for each route a own scope, e.g.

Ok, I don't see normal cases for collision, so $route should be ok.
But I don't want al-route makes a new scope, you can use al-ctrl if you want a new scope, also in this case you don't need `@route-to` because your controller be called when route is active.

**Comment by lega911** [comment]
$route is ready http://plnkr.co/edit/pQBycumCKE1DJb3oVaPE?p=preview

**Comment by lega911** [comment]
http://plnkr.co/edit/pQBycumCKE1DJb3oVaPE?p=preview
embedded: http://run.plnkr.co/plunks/pQBycumCKE1DJb3oVaPE/

**what works:**
al-route - define route (can be empty)
al-route:default - it's activated on undefined url (for 404)
al-route-out="onOut()" - return true if you want prevent url be changed
al-on.route-to - event when route is activated, you can use al-ctrl instead of it.
scope.$route - contains arguments from url

what else?

**Comment by lega911** [comment]
what about "href" ?

``` html
<a al-link="/login">Login</a>
<a al-router-link="/login">Login</a>
<a al-router-href="/login">Login</a>
```

Also we have js API:

``` js
service.location = {
    go(link)  // move to link
    subscribe(fn, [flag])  // subscribe on changes url, on 404 page, out-event
    unsubscribe(fn, [flag])
    getCurrentUrl()
    isDefault()  // returns true if it's 404
}
```

**Comment by rumkin** [comment]
What about routes that should not be routed and need page reloading? If I wish to serve `/blog` with another lib or have sub applications?

``` html
<ul al-router>
    <li al-route="/blog/**" al-route.reload></li>
</ul>
```

**Comment by lega911** [comment]
> need page reloading

you can use usual `<a href="">` to reload the page.

> If I wish to serve /blog with another lib or have sub applications?
> `al-route="/blog/**"`

Your example works, but router should be only one.
Your sub applications can use API: location.subscribe()

**Comment by rumkin** [comment]
> you can use usual`<a href="">` to reload the page.

ok

> what about "href" ?

al-href?

**Comment by lega911** [comment]
> al-href?

Why not, this name is not busy.

**Comment by fend25** [comment]
`al-href` will collide with `:href`, at least in mind.

AngularUI uses ui-sref attribute.

I vote for `al-router-link`. Although `al-router-href` is ok too.

**Comment by lega911** [comment]
vuejs and A2 seems to use name "link", v-link, router-link

**Comment by dev-rke** [comment]
Hm, your examples do not work anymore, maybe due to the last release?

**Comment by lega911** [comment]
It doesn't work in FF because of wrong transpiler, FF doesn't support sync xhr

**Comment by dev-rke** [comment]
I tried it in Chrome, it works fine.
I like the defined behaviour.

What do you think about prefixing normal href attributes using "#" instead of introducing something like al-link?
When a href attribute is prefixed with "#" the attribute gets handled by alight.
If html5 push states are used, the prefix get removed before applying the url.

**Comment by lega911** [comment]
Do you mean?

``` html
<a href="#/user/profile">Profile</a>
```

Interesting, what about buttons and other?

``` html
<button href="#/user/profile">Profile</button>
<div href="#/user/profile">Profile</div>
<li href="#/user/profile">Profile</li>
```

**Comment by rumkin** [comment]
Shebang is using to navigate across webview. This behaviour will be broken. I think `al-href` is good enough. It short and clean. It refers to default href which it actually is.

**Comment by fend25** [comment]
> What do you think about prefixing normal href attributes using "#" instead of introducing something like al-link?

it will bring some strange logic to the `service.go` processor.
But the most important thing is that it will collide and brake specs about hashes in url. And what if I want to use hashes? 

I think, special directive is the best way because it won't collide with specs nor with other things in mind. In appearance `al-href` is very similar to `:html`. And it hasn't any allusion to router. If nobody wants see `al-router-href`, maybe `al-rhref`? 

New idea: make new namespace for router: `r`, or `rt` or something else?

> Interesting, what about buttons and other?
> `<button href="#/user/profile">Profile</button>`

Why? it's totally confusing. And all workaround (like twitter bootstrap) is prepared for `<li><a></a></li>` structure.

**Comment by dev-rke** [comment]
@lega911

> Interesting, what about buttons and other?

href is meant as compatibility mode. So for using buttons normally one would use a link element and style it like a button (compare twitter bootstrap). Also a button can be used with a form, so the form action also should be handled.

I also think it would not be necessary to prefix links using a hash.
As we have a global routing object, we can check each click event of a link in the current scope, if the url matches to a defined route.
E.g.

``` html
<a href="/user">Will show /user route element</a>
<a href="/profile">There is no matching route, so page /profile will load</a>

<div al-route="/user"></div>
```

There should be no problems and it is easy to implement.
Also it would be good for SEO purposes, because one can use prerender.io and all links should be accessible.

@rumkin 

> Shebang is using to navigate across webview.

Can you please give an example?
I did not understand what the problem is. Maybe its obsolete when we check a clicked link against our routes (see above).

@fend25 

> But the most important thing is that it will collide and brake specs about hashes in url. And what if I want to use hashes? 

Why would it break? 
Using views requires to use hashes or html5 push states, because they are bound to the url.
When someone uses a link outside of defined routes, the link would not be handled by alight.
The great benefit would be, that someone can define "hard" links (e.g. for SEO purposes) which are handled by javascript, if the client supports javascript and fallback to normal links, if no javascript is used.

**Comment by lega911** [comment]
>  if the client supports javascript and fallback to normal links, if no javascript is used.

What if I need a normal link, but I use javascript.

> `<a href="/profile">There is no matching route, so page /profile will load</a>`

You can do this with al-route:default

**Comment by dev-rke** [comment]
> What if I need a normal link, but I use javascript.

Typically an url is unique (especially to prevent duplicate content in search engines).
So when defining a url that links to a view, the url is handled by alight.
When defining a url, that does not match a predefined route, the link will be handled by browser defaults (e.g. by opening the linked page).

> You can do this with al-route:default

I do not understand, what exactly you mean.

The concept is, that i define routes.
A route can be defined by al-route or by extending alight.routes object using custom javascript.
If a link (a tag with href attribute) is clicked, alight will try to match the href attribute to a route.
If a route matches the url, alight will handle the url (e.g. by showing a view) and prevent default browser behaviour.
If no route matches the url, alight will not intercept the default browser handling and the browser will handle the url, e.g. by loading the url.

**Comment by lega911** [comment]
> When defining a url, that does not match a predefined route

then "al-route:default" is activated

I think your approach can give us implicit mistakes

**Comment by lega911** [comment]
> I think your approach can give us implicit mistakes

e.g. if I open an application by wrong direct link, it would redirect me somewhere.

**Comment by dev-rke** [comment]
> then "al-route:default" is activated

When al-route:default is not defined, what happens then?
Nothing? I do not see the benefit of a "default" modifier.
If one needs a default setting, he can define a catchall route (e.g. "/.*") manually.
A default route should not be implicitly defined, it should be defined by the user.

Especially if you want to implement progressive enhancements (https://www.w3.org/wiki/Graceful_degradation_versus_progressive_enhancement#Graceful_degradation_and_progressive_enhancement_in_a_nutshell) on older websites, we need to ensure that the page can be used without modern features / excessive javascript usage.

> I think your approach can give us implicit mistakes

Can you please give an example?
Currently I do not see any case where there could be real problems.

The benefit is, that i do not need to define urls twice, when i want to make my application ready for SEO and for screenreaders.

Feature Request: defining a custom base path for the application, e.g. when an application is placed in a subfolder. A simple and not flexible solution would be to read the base tag from head element.
A better way would be to provide a custom interface, e.g. a property, which can be set manually.

**Comment by lega911** [comment]
> If one needs a default setting, he can define a catchall route (e.g. "/.*") manually.

This route will be visible always, route-default is visible only for undefined routes (page 404).

> the page can be used without modern features / excessive javascript usage.

I'm not care about old browsers for the router, because it's not a part of core (yet), and probably there will be a few variants of router.

> Can you please give an example?
> Maybe, if you make a misprint or if there is a mistake in code, the browser will reload page, and you will not see a error.

Also, how will you use "href" for buttons, it will look weird, http://getbootstrap.com/css/#buttons
So al-link/al-href should be anyway, your approach can be additional.

> defining a custom base path for the application

ok, it's should be useful.

**Comment by lega911** [comment]
> how will you use "href" for buttons, it will look weird

ok, for bootstrap you can use `a` for buttons, but I want it works for other elements, images/div/span...

**Comment by lega911** [comment]
> defining a custom base path for the application

done

``` js
service.location.setBase(location.pathname);
```

Added history-list example: http://plnkr.co/edit/pQBycumCKE1DJb3oVaPE?p=preview

**Comment by dev-rke** [comment]
> This route will be visible always, route-default is visible only for undefined routes (page 404).

So this is like a catchall.
Do you think this is necessary to provide in alight?
A user can define a catchall route manually.

> Also, how will you use "href" for buttons

I will NOT use the button tag. I will use an a tag and style it the same as a button.
This is what bootstrap does.
When using a button, the button relates to a form (e.g. as submit or reset button).
And the form needs to be also handled by alight, when a submit event occurs.

> but I want it works for other elements, images/div/span

Typically one defines an anchor element around an image or a span element.
A div should not be wrapped by an anchor, because a div is a block element while an anchor is an inline element. What you can do is using an anchor and a span and display them as block or inline-block elements.
So what you are thinking of here is to break pre-defined functionality. 
I think about merging existing standards with new functionality.
It also would be even simpler for new users just to define routes instead of changing the full template.
E.g. one already implemented a template and wants to enhance it with some client side features, it would be more simple to add alight when he does not need to modify existing links.
He just needs to define routes and a view which displays the contents.

**Comment by dev-rke** [comment]
Hm, what about this:

One defines his application e.g. /users/

``` html
<div al-route="/users/:id"></div>
<div al-route="/users/:id/social"></div>
<div al-route="/users/detail/:id/edit"></div>
<div al-route="/users">CatchAll for /users instead of al-route:default, e.g. when one removes paramters from the url</div>

<a href="/login">Redirects to a page /login</a>
<a href="/groups">Redirects to a page /groups</a>
```

This would be definitively more flexible than handling urls by using an extra directive.
It's interesting that you were against an own directive in combination with templates, but you prefer a own directive for routing. I am not against an own directive for routing, but i think it is not required.
It would be better to provide a simple API, e.g.:

``` html
<img @click="alight.go(/users/1234)" />
```

This would give more flexibility than a directive, because the directive is bound to a specific event, e.g. click event. But what when a view should be shown when something is hovered?

``` html
<img @mouseover="alight.go(/image/1234/large)" />
<div al-route="/image/:id/large">
  <img src="/resources/images/{{$route.id}}-large.jpg" />
</div>
```

This is a simple preview box of a large image, when the user hovers over a thumbnail image.
The great thing about this is, that each large image has an own url, so one can share this url. :-)

**Comment by dev-rke** [comment]
Better API proposal:

``` html
<img @click="$route.go(/users/1234)" />
```

**Comment by lega911** [comment]
> you were against an own directive in combination with templates, but you prefer a own directive for routing

Because template directives are all about one thing - inserting templates and I can reuse the code, no double code.
al-link has unique code, and it's shorter than al-route.link

> something is hovered?

a good idea, it's done

``` html
    <a @click="$route.go('/user/ubuntu/view')" href>User Ubuntu</a> <br/>
```

But you have a limitation here - you should use al-ctrl if you want to have "scope.$route"

**Comment by lega911** [comment]
> But you have a limitation here - you should use al-ctrl if you want to have "scope.$route"

I can place $route in every scope, but I think it's can be bothersome e.g. my timepicker component doesn't need it.
So $route is placed in child scope of al-route (al-route doesn't make a child scope), you can use al-ctrl to make a child scope where $route is appeared.

**Comment by lega911** [comment]
> `<div al-route="/users">CatchAll for /users instead of al-route:default`

"/user" - is a simple route, need a key-symbol, maybe "/user/*"?

**Comment by dev-rke** [comment]
> "/user" - is a simple route, need a key-symbol, maybe "/user/*"?

Nope, i explicitly mean /user.
This could be some kind of userlist or something, because it has no id in the url.
Matching against "/user/*" would exclude "/user", so it would not be a default handler.
If you append anything else, it would be matched by "/user/:id".

**Comment by lega911** [comment]
It seems we talk about different things, your route "/user" works.

**Comment by dev-rke** [comment]
> It seems we talk about different things

No, we talk about the same.
You want to implement a catchall handler, that catches all urls that are not matched by the given urls in a group, right.
I recommend to make the catchall configurable, by defining custom catchall handlers.
And i would like to see handling normal links.

What do you think about making the al-link directive as additional attribute, e.g. like:

``` html
<a href="/user/1234/detail" al-link>User 1234</a>
```

If al-link has no parameters, it tries to fetch the link from href attribute or from action attribute (for form elements) or from src attributes (for images). Is this a compromise? :-)

**Comment by lega911** [comment]
> CatchAll for /users instead of al-route:default, e.g. when one removes paramters from the url

I don't like this mark-way, what if I want "/user" as usuall route, it has no sign for this.
So, we can make special sign for this, we can make a special type for routes (e.g. :catch"),: 

``` html
<div al-route="/users/:id"></div>
<div al-route="/users/:id/social"></div>
<div al-route="/users/detail/:id/edit"></div>
<div al-route:catch="/users**">CatchAll for /users #1</div>
<div al-route:catch="/users**">CatchAll for /users #2</div>
<div al-route:catch="/**">CatchAll #3</div>
```

While al-route works in parallel (a few routes can be visible), al-route:catch depends on an order - only first al-route:catch will be visible, so for "/users/wrong/link/here" `CatchAll for /users #1` will be visible, for other urls it will be `CatchAll #3`

You can propose another sign for this

**Comment by dev-rke** [comment]
> I don't like this mark-way, what if I want "/user" as usuall route, it has no sign for this.

You can do this. The idea is _NOT_ to define a _DEFAULT_ route using a special directive or directive modifier.
You define your url handling, as it is done e.g. by express.
If you want to handle a 404, you need to specify on which route the 404 should be activated.
E.g. you define 

``` html
<div al-route="/users/details/:id"></div>
<div al-route="/users/details/:id/social"></div>
<div al-route="/users/details/:id/edit"></div>
<div al-route="/users">This is your /users route</div>
<div al-route="/users/.*">Catch, when no other route matches /users/(.*)</div>
```

Another example:

``` html
<div al-route="/community/users/:id"></div>
<div al-route="/community/users/:id/social"></div>
<div al-route="/community/users/detail/:id/edit"></div>
<div al-route="/community/:anything">This is the catcher, anything matches anything except users/*</div>
```

In this example any parameter, that does not match a given route, needs to be handled after other urls. We only need to define a processing priority, to ensure, that urls containing parameters are handled last. This can be done when adding new routes to the global routes object.

> While al-route works in parallel (a few routes can be visible), al-route:catch depends on an order - only first al-route:catch will be visible, so for "/users/wrong/link/here" `CatchAll for /users #1` will be visible, for other urls it will be `CatchAll #3`

I understand, what you like to do, but we do not need a directive modifier for this behaviour.
Its always the same handling structure.

**Comment by lega911** [comment]
> If you want to handle a 404, you need to specify on which route the 404 should be activated.
> E.g. you define
> 
> ``` html
> <div al-route="/users/details/:id"></div>
> <div al-route="/users/details/:id/social"></div>
> <div al-route="/users/details/:id/edit"></div>
> <div al-route="/users">This is your /users route</div>
> <div al-route="/users/.*">Catch, when no other route matches /users/(.*)</div>
> ```

It doesn't work now, because all routes work in parallel, e.g. if url = "/users/details/id123" then this DOM will be visible:

``` html
<div al-route="/users/details/:id"></div>
<div al-route="/users/.*">Catch, when no other route matches /users/(.*)</div>
```

> to ensure, that urls containing parameters are handled last

``` html
<div al-route="/users/details/details/:id"></div>
<div al-route="/users/details/:id/details"></div>
```

which one should be a last? (what if url = "/users/details/details/details")?

> I understand, what you like to do, but we do not need a directive modifier for this behaviour.
> Its always the same handling structure.

We have two modes for routes: parallel and sequence, and we need to know which  where.
Example:

``` html
<div al-route="/users/root"></div>

<div al-route:catch="/users/props/.*">props</div>
<div al-route:catch="/users/.*">Catch, when no other route matches /users/(.*)</div>
<div al-route:catch="/users/info/.*">info</div>  <!-- similar to props -->

<div al-route="/users/details/:id"></div>
<div al-route="/users/:id"></div>
```

For url=**"/users/root"**, result DOM: (It's Important case - two blocks are visible.)

``` html
<div al-route="/users/root"></div>

<div al-route="/users/:id"></div>  <!-- Double of route -->
```

url=**"/users/details/id123"**, result DOM:

``` html
<div al-route="/users/details/:id"></div>
```

url=**"/users/props/id123"**, result DOM:

``` html
<div al-route:catch="/users/props/.*">props</div>
```

url=**"/users/info/id123"**, result DOM:

``` html
<div al-route:catch="/users/.*">Catch, when no other route matches /users/(.*)</div>
```

It's Important case, you have `al-route:catch="/users/.*"` instead of al-route:catch="/users/info/.*", because of sequence (position in DOM)

What is your Skype? (mine: lega911), it will be easier ;)

**Comment by dev-rke** [comment]
> which one should be a last? (what if url = "/users/details/details/details")?

/users/details/:id
has precedence before
/users/:id/details

Why? Because static defined routes are more specific than routes containing a variable.
Also the segment length of a given route should be respected:

> It's Important case, you have `al-route:catch="/users/.*"` instead of `al-route:catch="/users/info/.*"`, because of sequence (position in DOM)

Your html:

``` html
<div al-route:catch="/users/.*">Catch, when no other route matches /users/(.*)</div>
<div al-route:catch="/users/info/.*">info</div>  <!-- similar to props -->
```

First `/users/info/.*` would be processed, afterwards `/users/.*`.
Why? Because `/users/info/.*` are three  segments `['users', 'info', '.*']` and `/users/.*` are only two segments `['users', '.*']`

We need to sort segments by length and we need to resolve position of all dynamic segments (variables like `.*` or `:id`).

> What is your Skype? (mine: lega911), it will be easier ;)

Talking wouldn't be that easy for me, because i need some time to think about what the best approach would be. If it is ok for you we can chat, but then we also could use the git chat application that you provided in another issue. :-)

**Comment by lega911** [comment]
> Talking wouldn't be that easy for me

I meant for chatting, it seems no one visits gitter

> First /users/info/.\* would be processed, afterwards /users/.*.

Ok, I see logic, but using the order of routes is easier, because you don't need to think of "static defined", count of segments etc.

More important - how can you do this:

> For url="/users/root", result DOM: (It's Important case - two blocks are visible.)
> 
> ``` html
> <div al-route="/users/root"></div>
> <div al-route="/users/:id"></div>  
> ```

These two blocks should be visible together for url "/users/root".
ngRoute has died because it had no this feauter, it had only one ngView

**Comment by dev-rke** [comment]
There are problems when adding a Route using JavaScript. So you need to sort the routes before processing them or it wont work.

To your second question: i see no case when two routes should be visible. There is exactly one visible DOM segment.

**Comment by lega911** [comment]
> To your second question: i see no case when two routes should be visible. There is exactly one visible DOM segment.

No, it exactly why ngRouter is dead, there are a lot of cases.
When you change url - top menu should be changed, left menu, body and footer. For this case you need use 4 similar routes, and they should be visible together.

**Comment by dev-rke** [comment]
Wouldn't it be more flexible to apply al-if instead of al-route?
E.g.:

``` html
<div al-if="$route.is('/users/:id')"></div>
```

But yes, i understand your case.

**Comment by lega911** [comment]
> Wouldn't it be more flexible

It's not easy to implement "confirm on exit" for this.
Also it is fired on every scan loop while al-route is working on url changing only (a better berformance).

I've also added al-route-out2 - it is fired on url chaning,
al-route-out is fired when you leave a "group"/route.

**Comment by dev-rke** [comment]
Wouldnt it be better to implement custom events?

``` html
<div @al.route.out="left()"></div>
```

**Comment by lega911** [comment]
It's good idea, I'll try it

``` html
<div @route-out="confirmToOut($url)"></div>
```

What about al-route and al-route:catch? Need good names for them.

**Comment by dev-rke** [comment]
> `<div @route-out="confirmToOut($url)"></div>`

I'd recommend to prefix framework specific events.
It's easier to understand what the code is about.

> What about al-route and al-route:catch? Need good names for them.

The problem is, it should not do a segment less catchall only (e.g. catching `/` for 404).
It only catches a specific route (e.g. to catch `/users` without parameters).
I accept al-route:catch.

We also need to discuss al-link vs. using the href attribute.
Would it be possible to enable href attributes using a config flag or by defining an empty al-link attribute?

**Comment by lega911** [comment]
> Would it be possible to enable href attributes using a config flag or by defining an empty al-link attribute?

yes

> it should not do a segment less catchall only (e.g. catching / for 404).

Do you mean, al-route:catch="/"  should catch all "/.*"?

**Comment by dev-rke** [comment]
> Do you mean, al-route:catch="/" should catch all "/.*"?

All, except already defined routes.
We do not need to define variable segments using `.*`, because it is a "catcher", it has to catch all urls that do not match any defined route.
The good thing is, that we do not need to process `.*` in any url.
We can always force the user to define a variable name, e.g. "/:id" or "/:anything".

**Comment by dev-rke** [comment]
I'd like to push on this, because it will be a great feature.
It is also ok when the first implementation hasn't a final API, because we can introduce this as not stable feature.
The API get's better and better when we can use the unstable API for development and test purposes, 
this will show up which other feature are required for productive usage.

Do you need any further information to implement this?
Or do you like to focus on other features? 
Then i would implement a quick proposal.

**Comment by lega911** [comment]
I will append it to package "ext"

**Comment by dev-rke** [comment]
Where will it be found? In the regular repository?

**Comment by lega911** [comment]
> In the regular repository?

Yes

**Comment by dev-rke** [comment]
Is it ready? I'd like to play with it. :-)

**Comment by lega911** [comment]
Hello! I was quite busy, published it here: https://github.com/lega911/alight/blob/master/ext/router.es.js
source here: https://github.com/lega911/angular-light/blob/master/src/ext/router.es.js
direct link: https://rawgit.com/lega911/alight/master/ext/router.es.js
example: http://plnkr.co/edit/HpUDXdTJ0QFenprkE5QJ?p=preview

**Comment by dev-rke** [comment]
Great to have you back. :-)
Is it contained in the bower package?
I'll give it a try tomorrow :-)

**Comment by lega911** [comment]
> Is it contained in the bower package?

Not yet, but it will be there in next version
